psygnal/_evented_model.py

```diff
@@ -0,0 +1,456 @@
+                    cls.__property_setters__[name] = attr
+                    signals[name] = Signal(object)
+
+        cls.__field_dependents__ = _get_field_dependents(cls)
+        cls.__signal_group__ = type(f"{name}SignalGroup", (SignalGroup,), signals)
+        return cls
+
+
+def _get_field_dependents(cls: EventedModel) -> Dict[str, Set[str]]:
+    """Return mapping of field name -> dependent set of property names.
+
+    Dependencies may be declared in the Model Config to emit an event
+    for a computed property when a model field that it depends on changes
+    e.g.  (@property 'c' depends on model fields 'a' and 'b')
+
+    Examples
+    --------
+        class MyModel(EventedModel):
+            a: int = 1
+            b: int = 1
+
+            @property
+            def c(self) -> List[int]:
+                return [self.a, self.b]
+
+            @c.setter
+            def c(self, val: Sequence[int]):
+                self.a, self.b = val
+
+            class Config:
+                property_dependencies={'c': ['a', 'b']}
+    """
+    deps: Dict[str, Set[str]] = {}
+
+    cfg_deps = getattr(cls.__config__, PROPERTY_DEPENDENCIES, {})  # sourcery skip
+    if cfg_deps:
+        assert isinstance(cfg_deps, dict), "Config.property_dependencies must be a dict"
+        for prop, fields in cfg_deps.items():
+            if prop not in cls.__property_setters__:
+                raise ValueError(
+                    "Fields with dependencies must be property.setters."
+                    f"{prop!r} is not."
+                )
+            for field in fields:
+                if field not in cls.__fields__:
+                    warnings.warn(f"Unrecognized field dependency: {field!r}")
+                deps.setdefault(field, set()).add(prop)
+    if getattr(cls.__config__, GUESS_PROPERTY_DEPENDENCIES, False):
+        # if property_dependencies haven't been explicitly defined, we can glean
+        # them from the property.fget code object:
+        # SKIP THIS MAGIC FOR NOW?
+        for prop, setter in cls.__property_setters__.items():
+            if setter.fget is not None:
+                for name in setter.fget.__code__.co_names:
+                    if name in cls.__fields__:
+                        deps.setdefault(name, set()).add(prop)
+    return deps
+
+
+class EventedModel(BaseModel, metaclass=EventedMetaclass):
+    """A pydantic BaseModel that emits a signal whenever a field value is changed.
+
+    In addition to standard pydantic `BaseModel` properties (see
+    https://pydantic-docs.helpmanual.io/usage/models/), this class adds the following:
+
+    1. gains an `events` attribute that is an instance of `psygnal.SignalGroup`. This
+       group will have a signal for each field in the model (excluding private
+       attributes and non-mutable fields).  Whenever a field in the model is mutated,
+       the corresponding signal will emit with the new value (see example below).
```

Comment:
we have objects in napari which provide multiple events (`changing`, `changed`) per 'evented' attribute -> is there some reason we don't need these now? I'm wondering if anyone will feel limited by the event only happening post value change (computing deltas might be hard?)

Definitely not suggesting this is necessary right now, just curious

Reply:
The only reason I did that (in the evented list) was because qt needs to know both before and after that something is changing... mostly so that it can call [`beginMoveRows`](https://doc.qt.io/qt-5/qabstractitemmodel.html#beginMoveRows) beforehand, and [`endMoveRows`](https://doc.qt.io/qt-5/qabstractitemmodel.html#endMoveRows) at the end.   I don't particularly love having all those events otherwise :) so i guess I'd wait until something else needs that before implementing globally